nomt/src/seglog/segment_rw.rs

```diff
@@ -0,0 +1,462 @@
+    }
+}
+
+pub struct SegmentFileReader {
+    buf_reader: BufReader<File>,
+    header: [u8; HEADER_SIZE as usize],
+    /// The position of the next record.
+    ///
+    /// `None` if the header has not been read yet.
+    next_pos: Option<u64>,
+    /// The payload length of the current record.
+    ///
+    /// `None` if the header has not been read yet.
+    payload_length: Option<u32>,
+    /// The size of the file.
+    file_size: u64,
+}
+
+impl SegmentFileReader {
+    pub fn new(file: File, file_size: Option<u64>) -> Result<Self> {
+        let file_size = if let Some(file_size) = file_size {
+            file_size
+        } else {
+            file.metadata()?.len()
+        };
+        Ok(Self {
+            buf_reader: BufReader::new(file),
+            header: [0u8; HEADER_SIZE as usize],
+            next_pos: None,
+            payload_length: None,
+            file_size,
+        })
+    }
+
+    /// Reads the header of the record.
+    ///
+    /// Returns `None` if the end of the file is reached.
+    pub fn read_header(&mut self) -> Result<Option<RecordHeader>> {
+        if self.next_pos.unwrap_or(0) >= self.file_size {
+            return Ok(None);
+        }
+        self.buf_reader.read_exact(&mut self.header)?;
+        let header = RecordHeader::new(&self.header);
+        if header.payload_length() > MAX_RECORD_PAYLOAD_SIZE {
+            // Rewind the buffer reader to the beginning of the header so that the next read
+            // will read the next header.
+            self.buf_reader
+                .seek(SeekFrom::Current(-(HEADER_SIZE as i64)))?;
+            return Err(anyhow::anyhow!(
+                "Record payload length is too large: {}",
+                header.payload_length()
+            ));
+        }
+        self.payload_length = Some(header.payload_length());
+        // It's not enough to just seek to the payload length because the records always span
+        // an integer number of pages.
+        let cur_pos = self.buf_reader.stream_position()?;
+        let record_alignment = RECORD_ALIGNMENT as u64;
+        let next_pos = ((cur_pos + self.payload_length.unwrap() as u64 + record_alignment - 1)
+            / record_alignment)
+            * record_alignment;
+        self.next_pos = Some(next_pos);
+        Ok(Some(header))
+    }
+
+    /// Skips the payload of the record.
+    ///
+    /// Must be called after `read_header`.
+    pub fn skip_payload(&mut self) -> Result<()> {
+        self.seek_next()?;
+        Ok(())
+    }
+
+    /// Reads the payload of the record.
+    ///
+    /// Must be called after `read_header`.
+    pub fn read_payload(&mut self, payload_buf: &mut Vec<u8>) -> Result<()> {
+        payload_buf.resize(self.payload_length.unwrap() as usize, 0);
+        self.buf_reader.read_exact(payload_buf)?;
+        self.seek_next()?;
+        Ok(())
+    }
+
+    /// Positions the reader at the beginning of the next record.
+    pub fn seek_next(&mut self) -> Result<()> {
```

Comment:
worth adding a comment (here and elsewhere) that usage panics when the header hasn't yet been read.

Reply:
check v3